net: wifi: esp32: update driver to include wifi API calls

sylvioalves opened the pull request:
This PR enables ESP32 driver to work with Zephyr's wifi API mgmt.

- Running `samples/net/wifi` using a ESP32 board is now possible, which allows WiFi scanning and connect/disconnect calls.  
- A few other drivers and readme files were changed as well as a result of the above changes.
- `samples/boards/esp32/wifi_station` was delete as it no longer works with this change.
- Added a customized Kconfig for ESP32 board to allow automatic DHCP bound.

sylvioalves committed:
wifi: esp32: add support to wifi api mgmt

Moved all MBEDTLS dependencies from prj.conf
to Kconfig as WiFi depends on it.

Update esp32 wifi driver to enable `samples/net/wifi`
to work. Commands as such as `wifi connect` and `wifi scan` are now
available.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
wifi: esp32: add config to enable automatic DHCPV4 negotiation

ESP32 wifi connection uses Zephyr's net stack. Once WiFi connects
in station mode, is requires the application to handle DHCPv4
negotiation. This PR adds support to automatic negotiation by
handling the dhcpv4 calls in driver layer.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
samples: boards: remove esp32 wifi sample code

`samples/net/wifi` replaces custom esp32 wifi sample
code. Then it can be removed from main repository.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

rlubos commented on samples/net/wifi/wifi_shell/README.rst:
```diff
@@ -1,7 +1,7 @@
-.. _wifi_sample:
+.. _wifi_shell_sample:
```
I'm not able comment on the sample's CMakeLists.txt, but you should also update the project name in there.

rlubos commented on samples/net/wifi/wifi_station/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(wifi)
```
`wifi_station`

rlubos commented on samples/net/wifi/wifi_station/src/main.c:
```diff
@@ -0,0 +1,76 @@
+			break;
+		}
+
+		LOG_INF("Connect request failed %d. Waiting iface be up...", ret);
```
The sample could register for `NET_EVENT_IF_UP` event I think before proceeding with the connection attempt?

mniestroj commented on drivers/wifi/Kconfig:
```diff
@@ -29,6 +29,16 @@ config WIFI_OFFLOAD
 	help
 	  Enable support for Full-MAC Wi-Fi devices.
 
+config WIFI_SSID
+	string "SSID of WiFi network"
+	help
+	  SSID (network name) for the application to connect to.
+
+config WIFI_PSK
+	string "Password of WiFi network"
+	help
+	  WiFi password for the application.
+
```
I agree with @rlubos, this should be part of sample configuration only.

sylvioalves replied:
ok!

mniestroj commented on include/zephyr/net/wifi.h:
```diff
@@ -24,4 +26,10 @@ enum wifi_security_type {
 #define WIFI_CHANNEL_MAX 14
 #define WIFI_CHANNEL_ANY 255
 
+struct wifi_api {
+	struct ethernet_api eth_api;
+
+	/* Any WiFi specific functionality can be added here. */
+};
+
```
This pulls in ethernet_api structure to all drivers, even tough most of them (all besides ESP32 as of now) do not (explicitly, at least to Zephyr knowledge) operate on ethernet frames at all.

sylvioalves replied:
@mniestroj My initial PR was using ifdefs but I get @rlubos idea. Any suggestion regarding this so I can update this?

sylvioalves replied:
Hi, any comments?

rlubos replied:
@sylvioalves Can you elaborate on this topic a bit?

I recall that when we initially discussed native WiFi support in Zephyr it was said that WiFi is compatible with Ethernet at L2 level? That's why we decided to reuse the existing Ethernet L2, instead of implementing a separate L2 for WiFi, which would be way more complex task. I'm not expert in the area, but for sure this approach works here for Espressiff as well as for Nordic devices, where we use the same approach of reusing Ethernet L2.

sylvioalves replied:
@mniestroj @rlubos It works with Ethernet L2. For now, there is no plan of changing this. I do agree we should add in Zephyr's roadmap `WiFi native` support, which means among other implementations, adding wpa_supplicant support.  
While we don't have it, we can proceed choosing the options discussed before:
1) Use `#ifdef` to allow Ethernet L2 + WiFi struct, which won't allow simultaneous `wifi offloaded` and `non-offloaded` usage, which is pretty much fine by me.
2) Keep as is and increase resource usage of `wifi offloaded` implementation.
@rlubos, Do you see any alternative like creating a new `wifi_native` interface, which won't touch `net_wifi_mgmt_offload` and its related stuff?

rlubos replied:
Defining an alternative interface structure is also an option, although as far as I understand, it would still need to expose the offloaded mgmt API? Perhaps we should extract it into a separate struct that would be then included by both interface APIs?

sylvioalves replied:
So, I have added the `#ifdef` again in there. I honest think we can start with this approach and update in the future. This won't allow both OFFLOADED and NON-OFFLOADED usage. Please, take a look @mniestroj @rlubos.

rlubos replied:
ok, I guess we can live with that for the time being.

mniestroj replied:
I think that we confuse socket/networking offloading (like send/recv, TCP/IP stack) with WiFi supplicant offloading (like doing scans, connections, managing WPA, etc.) in this thread. What I still don't buy is including `struct ethernet_api` inside `struct net_wifi_mgmt`. What is the rationale behind it? ESP32 is implementing single function from `struct ethernet_api`, which is `.send()`. Such function exists also in `struct ppp_api` or `struct virtual_interface_api`. So why is it Ethernet and not PPP or Virtual that ESP32 WiFi driver implements?

Looking further into the driver, it seems that the reason why Ethernet is used for ESP32 WiFi is because `esp_wifi_internal_tx()` and `esp_wifi_internal_reg_rxcb()` expect Ethernet frames (i.e. payload prepended with DST, SRC, TYPE/LEN fields). Okay then, let that be, that is what it is **for that driver**. But is that enough reason to actually add `struct ethernet_api` into `struct net_wifi_mgmt`? IMO it is not.

I had a look also into nrf700x driver in NCS repository. It also relies on Ethernet frames for RX/TX in https://github.com/nrfconnect/sdk-nrf/blob/main/drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c. But why is that? Because it tries to pretend Ethernet device (same as ESP32 WiFi), by converting WiFi (802.11) frames into Ethernet frames.

Both ESP32 and nrf700x drivers are operating on Ethernet frames. To me it looks like a workaround, not something that comes from WiFi specification or will be a common thing for WiFi drivers. So instead of having:
```
#ifdef CONFIG_WIFI_OFFLOAD
	struct net_if_api wifi_iface;
#else
	struct ethernet_api wifi_iface;
#endif
```
which suggest that WiFi offloaded driver are a special case, I would propose:
```
#ifdef CONFIG_WIFI_ESP32
	struct ethernet_api wifi_iface;
#else
	struct net_if_api wifi_iface;
#endif
```
which is a workaround for ESP32 driver trying to emulate Ethernet.

sylvioalves replied:
@rlubos, any comments regarding the above?  
@mniestroj I get your point regarding that it is specific do ESP32 in this case.  
Should we use more generic approach like `#ifdef CONFIG_WIFI_OVER_ETHERNET_IFACE` for that instead of having custom vendor #ifdef within wifi_mgmt sources?

mniestroj replied:
>Should we use more generic approach like #ifdef CONFIG_WIFI_OVER_ETHERNET_IFACE for that instead of having custom vendor #ifdef within wifi_mgmt sources?

That sounds good to me! Probably nrf700x would use the same in the early days.

sylvioalves replied:
@rlubos, would you review again?

sylvioalves replied:
Added the `CONFIG_WIFI_OVER_ETHERNET_IFACE` approach.

krish2718 replied:
> Both ESP32 and nrf700x drivers are operating on Ethernet frames. To me it looks like a workaround, not something that comes from WiFi specification or will be a common thing for WiFi drivers.

For FullMAC chips we can pass data frames in any format, its mostly `payload + meta_data`, and the UMAC in the chip uses the meta_data to build 802.11 header a.k.a encapsulation. The general norm here is instead of each vendor defining their own meta_data, Ethernet header is used as meta data as it has all necessary information for the UMAC to build 802.11 header. So, even in Linux all FullMAC Wi-Fi chipsets are essentially treated as Ethernet devices in data path.

Wi-Fi chips in Zephyr are doing below, this is not a hack, we definitely need an option for this either default or using `CONFIG_WIFI_OVER_ETHERNET_IFACE` to save some memory.

* esp32 - (FullMAC + 802.1x supplicant)
    - Ethernet APIs + Kconfig based Wi-Fi configuration
* All others (FullMAC + 802.1x supplicant + TCP/IP)
    - wifi_mgmt + WIFI_OFFLOAD (NET_OFFLOAD a.k.a TCP/IP stack offload)
* nrf700x - (only FullMAC)
    -  wifi_mgmt + Native TCP/IP

mniestroj replied:
Thanks @krish2718 for your input!

>So, even in Linux all FullMAC Wi-Fi chipsets are essentially treated as Ethernet devices in data path.

Just curious, could you point to some example of that? I was taking a quick look myself, but couldn't find anything useful in Linux tree that we can follow. I just don't have much WiFi experience, so that might be the case.

mniestroj commented on samples/boards/esp32/wifi_station/src/main.c:
```diff
@@ -1,81 +0,0 @@
-	if (!IS_ENABLED(CONFIG_ESP32_WIFI_STA_AUTO)) {
-		wifi_config_t wifi_config = {
-			.sta = {
-				.ssid = CONFIG_ESP32_WIFI_SSID,
-				.password = CONFIG_ESP32_WIFI_PASSWORD,
```
Those config options should be removed from Kconfig.

mniestroj commented on samples/net/wifi/wifi_station/src/main.c:
```diff
@@ -0,0 +1,76 @@
+	cnx_params.ssid_length = strlen(CONFIG_WIFI_SSID);
+	cnx_params.psk_length = strlen(CONFIG_WIFI_PSK);
+
+	LOG_INF("connecting to %s...", CONFIG_WIFI_SSID);
```
Not sure what is wrong, but I couldn't connect to any of my 2 APs that I have at home. It just hangs at:
```
[00:00:00.826,000] <inf> wifi_station: connecting to AX3600...
```
Didn't try to debug yet.

rlubos reviewed (dismissed):
LGTM

mniestroj commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -116,14 +117,18 @@ struct wifi_status {
 	 * that can be casted to a net_if_api structure pointer.
 	 */
-	struct net_if_api iface_api;
+#ifdef CONFIG_WIFI_OFFLOAD
+	struct net_if_api wifi_iface;
+#else
+	struct ethernet_api wifi_iface;
+#endif
```
It is not that CONFIG_WIFI_OFFLOAD are anything special. It is ESP32 that operates on Ethernet and requires to embed `struct ethernet_api`.

loicpoulain replied:
You can probably use union here.

sylvioalves commented:
@krish2718 @rlubos, would you check the latest update regarding the kconfig option?

```c
config WIFI_OVER_ETHERNET_IFACE
	bool "Support Wi-Fi device drivers over ethernet interface"
	help
	  Enable wifi driver to use ethernet API.
```

krish2718 commented:
oops, didn't check the existing PRs, but my PR is a subset of this https://github.com/zephyrproject-rtos/zephyr/pull/49312/files so, I should probably drop it.

krish2718 commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -194,45 +459,39 @@ static int eth_esp32_dev_init(const struct device *dev)
-	.send =  eth_esp32_send,
+static const struct net_wifi_mgmt_offload esp32_api = {
+	.wifi_iface.iface_api.init = esp32_wifi_init,
+	.wifi_iface.send = esp32_wifi_send,
```
shouldn't this also be protected with CONFIG_WIFI_OVER_ETHERNET_IFACE ?

sylvioalves replied:
Well, `esp_wifi_drv.c` is only added into build when ESP32+WIFI is select. Otherwise, I see no need to protect it. Do you?

krish2718 replied:
Ah that is auto-selected in Kconfig, should be fine, if someone removes that it will throw build error anyway.

krish2718 commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -194,45 +459,39 @@ static int eth_esp32_dev_init(const struct device *dev)
 #if defined(CONFIG_NET_STATISTICS_ETHERNET)
-	.get_stats = eth_esp32_stats,
-#endif
+	.wifi_iface.get_stats = esp32_wifi_stats,
```
If you really want wifi_stats, then please see my other PR https://github.com/zephyrproject-rtos/zephyr/pull/49049/files#diff-d669b57333cbd97e6ffe0af5de4e493c5c459de1d59ad8cf0fa5ece3a6e4cf0fR489 I have only added bare minimum to start with but we can easily extend.

sylvioalves replied:
I'll keep this as is and update once the mentioned PR gets merged, sounds good?

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -148,15 +153,11 @@ struct net_wifi_mgmt_offload {
 void wifi_mgmt_raise_connect_result_event(struct net_if *iface, int status);
 void wifi_mgmt_raise_disconnect_result_event(struct net_if *iface, int status);
 
-#endif /* CONFIG_WIFI_OFFLOAD */
```
Urgh..I did a similar change here https://github.com/zephyrproject-rtos/zephyr/pull/49049/commits/31c8b4fc7a69256affcc96f1cee22314a9bb9405

krish2718 commented on samples/net/wifi/wifi_station/src/main.c:
```diff
@@ -0,0 +1,80 @@
+/*
```
I guess `subsys/net/l2/wifi_shell.c` is a superset, right, we don't need a separate wifi_shell sample?

sylvioalves replied:
You mean removing `wifi_shell` sample?

krish2718 replied:
Not quite, I don't see what extra this sample is adding that is not there in the `wifi_shell` sample? Except that `wifi_shell` automatically selects `wifi_mgmt` but I guess that is fine.

sylvioalves replied:
I see. The whole idea is to have a `station` sample and not rely on the `wifi_shell.c` subsystem. I find it very difficult for new users to get Zephyr's working way without sample codes. Anyway, `station` sample is just duplicating, I can removed it. I'll wait more feedback.

krish2718 replied:
The commit log needs to change accordingly.

krish2718 commented:
> ```c
> WIFI_OVER_ETHERNET_IFACE
> ```
nit, `WIFI_WITH_NATIVE_TCP_IP` or `WIFI_WITH_NATIVE_NETWORKING` might be more appropriate, `OVER` sounds like some kind of tunnelling (at least to me).

loicpoulain commented:
> > ```c
> > WIFI_OVER_ETHERNET_IFACE
> > ```
> 
> nit, `WIFI_WITH_NATIVE_TCP_IP` or `WIFI_WITH_NATIVE_NETWORKING` might be more appropriate, `OVER` sounds like some kind of tunnelling (at least to me).

I agree, it's not wifi over ethernet here, It's just that the MAC layer is offloaded to the controller and ethernet frames are used for data path. maybe something like WIFI_DATA_8023 or WIFI_IFACE_ETH...

sylvioalves commented:
> WIFI_WITH_NATIVE_NETWORKING

Sure, It make sense.  
`WIFI_WITH_NATIVE_NETWORKING`  
`WIFI_USE_NATIVE_NETWORKING`  (I prefer this)
`WIFI_IFACE_ETH`  

@mniestroj @rlubos PTAL.

loicpoulain reviewed (commented):
Maybe rename/refactor 'station' to 'client' or 'client station'. A WiFi AP is also a station, and the sample is dedicated to client mode, right?

sylvioalves commented:
> Maybe rename/refactor 'station' to 'client' or 'client station'. A WiFi AP is also a station, and the sample is dedicated to client mode, right?

I'll rename to `station` then.

krish2718 commented:
> WIFI_USE_NATIVE_NETWORKING

Fine with me.

krish2718 commented:
> Naming "offload" in "struct net_wifi_mgmt_offload" is kept because
Zephyr still has no supplicant to handle a full non-offloaded driver.

FYI, Zephyr doesn't but `sdk-nrf` does and in future it will be upstreamed to Zephyr as well.

krish2718 commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -44,6 +56,22 @@ K_MSGQ_DEFINE(esp_wifi_msgq, sizeof(system_event_t), 10, 4);
+
+	switch (mgmt_event) {
+	case NET_EVENT_IPV4_DHCP_BOUND:
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface, 0);
```
Typically, this is only for L2 connected, so, if we delay/fail this in case DHCP delays/fails, then applications behave differently with ESP32 vs. other Wi-Fi chips, is that okay? Ideally DHCP restart/start should be handled by stack/apps higher up. In `nrf700x` DHCP restart is handled by wpa_supplicant, FYI here also a new API was added https://github.com/zephyrproject-rtos/zephyr/pull/49152

sylvioalves replied:
Sure, I get that. And this is only used when `CONFIG_ESP32_WIFI_STA_AUTO_DHCPV4` is selected. By default, DHCP should be handled by the app. Thanks for pointing the DHCP restar PR.

jukkar commented:
> `WIFI_USE_NATIVE_NETWORKING` (I prefer this)

:+1: I agree, this one looks best of these.

mniestroj commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -119,32 +128,101 @@ static esp_err_t eth_esp32_rx(void *buffer, uint16_t len, void *eb)
+			if (esp32_data.scan_cb) {
+				esp32_data.scan_cb(esp32_data.iface, 0, &res);
+
+				/* ensure notifications get delivered */
+				k_yield();
```
This is just a workaround IMO, it doesn't ensure anything.

mniestroj commented on drivers/wifi/esp32/Kconfig.esp32:
```diff
@@ -5,6 +5,10 @@ menuconfig WIFI_ESP32
+	select MBEDTLS
+	select MBEDTLS_ENTROPY_ENABLED
+	select MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
+	select MBEDTLS_ECP_ALL_ENABLED
```
Really, all ECP are needed?

mniestroj replied:
Also: could you clarify in commit message why are you pulling those mbedTLS dependencies? It is not clear to me.

sylvioalves replied:
@mniestroj I will add this info into commit message. I just moved `MBEDTLS*` from `prj.conf` as this is mandatory dependency. https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/esp32/wifi_station/prj.conf#L17  
I'll work you ECP dependencies in another PR. For now, just keeping what is already upstreamed.

mniestroj commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -153,37 +231,181 @@ static void esp_wifi_event_task(void)
+	wifi_config.sta.pmf_cfg.capable = true;
+	wifi_config.sta.pmf_cfg.required = false;
+
+	ret = esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config);
+	ret |= esp_wifi_set_mode(WIFI_MODE_STA);
+	ret |= esp_wifi_connect();
+
+	return ret;
```
We should return a posix error code, not just some integer.

mniestroj commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -153,37 +231,181 @@ static void esp_wifi_event_task(void)
+	}
+
+	/* Start WiFi in AP mode with configuration built above */
+	esp_err_t err = esp_wifi_set_mode(WIFI_MODE_AP);
+
+	if (err != ESP_OK) {
+		LOG_ERR("Failed to set WiFi mode: %d", err);
+		return err;
+	}
+	err = esp_wifi_set_config(WIFI_IF_AP, &wifi_config);
+	if (err != ESP_OK) {
+		LOG_ERR("Failed to set WiFi config: %d", err);
+		return err;
+	}
+	err = esp_wifi_start();
+	if (err != ESP_OK) {
+		LOG_ERR("Failed to start WiFi: %d", err);
+		return err;
```
Error codes should be mapped to posix error codes, as application layer will depend on those.

mniestroj commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -198,41 +420,31 @@ static int eth_esp32_dev_init(const struct device *dev)
 	wifi_init_config_t config = WIFI_INIT_CONFIG_DEFAULT();
 	esp_err_t ret = esp_wifi_init(&config);
 
-	ret |= esp_supplicant_init();
+	ret |= esp_wifi_set_mode(WIFI_MODE_STA);
 	ret |= esp_wifi_start();
 
-	if (IS_ENABLED(CONFIG_ESP32_WIFI_STA_AUTO)) {
-		wifi_config_t wifi_config = {
-			.sta = {
-				.ssid = CONFIG_ESP32_WIFI_SSID,
-				.password = CONFIG_ESP32_WIFI_PASSWORD,
-			},
-		};
-
-		ret = esp_wifi_set_mode(WIFI_MODE_STA);
-		ret |= esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config);
-		ret |= esp_wifi_connect();
-	}
-
 	if (ret != ESP_OK) {
-		LOG_ERR("Connect failed");
+		LOG_ERR("Wifi driver failed");
```
Same here, posix error codes are expected in general in Zephyr, so please convert those somehow before returning.

mniestroj commented on samples/net/wifi/boards/esp32.conf:
```diff
@@ -0,0 +1,18 @@
+CONFIG_NET_LOG=y
+
+CONFIG_ESP32_WIFI_SSID="ssid"
+CONFIG_ESP32_WIFI_PASSWORD="password"
```
Shouldn't ssid and password be passed to "connect" wifi_mgmt API? Why are those needed here?

sylvioalves replied:
Not needed, removing.

mniestroj commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -153,37 +231,181 @@ static void esp_wifi_event_task(void)
+
+	wifi_scan_config_t scan_config = { 0 };
+
+	ret = esp_wifi_set_mode(WIFI_MODE_STA);
+	ret |= esp_wifi_scan_start(&scan_config, false);
+
+	if (ret != ESP_OK) {
+		return -EIO;
+	}
```
Posix error codes, please. Same as in other places.

mniestroj commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -198,41 +420,31 @@ static int eth_esp32_dev_init(const struct device *dev)
+	ret |= esp_wifi_set_mode(WIFI_MODE_STA);
 	ret |= esp_wifi_start();
 
-	if (IS_ENABLED(CONFIG_ESP32_WIFI_STA_AUTO)) {
```
Kconfig option should be removed with the same commit. I guess that `CONFIG_ESP32_WIFI_SSID` should be removed as well, or is it still used somewhere?

sylvioalves replied:
`CONFIG_ESP32_WIFI_SSID` removed.

mniestroj commented on samples/net/wifi/boards/esp32c3_devkitm.conf:
```diff
@@ -0,0 +1,18 @@
+CONFIG_NET_LOG=y
+
+CONFIG_ESP32_WIFI_STA_AUTO_DHCPV4=y
+CONFIG_ESP32_WIFI_STA_RECONNECT=n
```
Maybe it should be disabled by default in Kconfig, so that application layer with use of wifi_mgmt API will take care of reconnection?

mniestroj commented:
>>WIFI_USE_NATIVE_NETWORKING (I prefer this)
>
>+1 I agree, this one looks best of these.

:+1: . @sylvioalves would you create a PR with that single commit, so @krish2718 can progress with his work that depends on that? We can review ESP32 specific part separately then.

krish2718 commented:
> > > WIFI_USE_NATIVE_NETWORKING (I prefer this)
> > 
> > 
> > +1 I agree, this one looks best of these.
> 
> +1 . @sylvioalves would you create a PR with that single commit, so @krish2718 can progress with his work that depends on that? We can review ESP32 specific part separately then.

I was about to suggest this myself, thanks @mniestroj .

sylvioalves commented:
> > > > WIFI_USE_NATIVE_NETWORKING (I prefer this)
> > > 
> > > 
> > > +1 I agree, this one looks best of these.
> > 
> > 
> > +1 . @sylvioalves would you create a PR with that single commit, so @krish2718 can progress with his work that depends on that? We can review ESP32 specific part separately then.
> 
> I was about to suggest this myself, thanks @mniestroj .

Done. Will update this PR soon.

sylvioalves committed:
wifi: esp32: update enum naming reference

Due to Zephyr's WiFi updates, common private
enums used in `hal_espressif` are now being used.

This updates private internal values to meet hal_espressif
changes.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented:
Added DNM label as this depends on hal_v4.4.1 updates provided from #49682.

krish2718 commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -153,37 +274,174 @@ static void esp_wifi_event_task(void)
+		wifi_config.sta.password[params->psk_length] = '\0';
+		wifi_config.sta.threshold.authmode = WIFI_AUTH_WPA2_PSK;
+	} else {
+		wifi_config.sta.threshold.authmode = WIFI_AUTH_OPEN;
```
with latest changes that extend the params->security, this is not true, so, need to add an explicit security check.

sylvioalves replied:
@krish2718, that's true. I'll add those new securities in the loop.

krish2718 commented on samples/net/wifi/boards/esp32.conf:
```diff
@@ -0,0 +1,14 @@
+CONFIG_HEAP_MEM_POOL_SIZE=98304
+
+CONFIG_NETWORKING=y
+CONFIG_NET_L2_ETHERNET=y
```
Shouldn't this be CONFIG_NET_L2_WIFI_MGMT?

sylvioalves replied:
No, it still uses L2_ETHERNET layer.

krish2718 commented on samples/net/wifi/boards/esp32.conf:
```diff
@@ -0,0 +1,14 @@
+CONFIG_NET_IPV4=y
+CONFIG_NET_DHCPV4=y
+
+CONFIG_NET_MGMT=y
+CONFIG_NET_MGMT_EVENT=y
+
```
CONFIG_NET_L2_WIFI_MGMT automatically chooses these

fabiobaltieri commented:
Can you rebase please?

sylvioalves commented:
> Can you rebase please?

done!

Machine-Hum commented:
Thanks for this work. I'm getting some linking errors 

```
CONFIG_WIFI=y
CONFIG_HEAP_MEM_POOL_SIZE=98304

CONFIG_NETWORKING=y
CONFIG_NET_L2_ETHERNET=y

CONFIG_NET_IPV6=n
CONFIG_NET_IPV4=y
CONFIG_NET_DHCPV4=y

CONFIG_NET_MGMT=y
CONFIG_NET_MGMT_EVENT=y

CONFIG_NET_LOG=y
CONFIG_NET_SHELL=n

CONFIG_MBEDTLS=y
CONFIG_MBEDTLS_ENTROPY_ENABLED=y
CONFIG_MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED=y
CONFIG_MBEDTLS_ECP_ALL_ENABLED=y

CONFIG_ESP32_WIFI_SSID=
CONFIG_ESP32_WIFI_PASSWORD=
CONFIG_GPIO=y

CONFIG_I2C=y

CONFIG_HEAP_MEM_POOL_SIZE=37760

# when enabling NET_SHELL, the following
# helps to optimize memory footprint
CONFIG_ESP32_WIFI_STATIC_RX_BUFFER_NUM=8
CONFIG_ESP32_WIFI_DYNAMIC_RX_BUFFER_NUM=8
CONFIG_ESP32_WIFI_DYNAMIC_TX_BUFFER_NUM=8
CONFIG_ESP32_WIFI_IRAM_OPT=n
CONFIG_ESP32_WIFI_RX_IRAM_OPT=n

# Include IQS driver
CONFIG_SENSOR=y
CONFIG_IQS5XX=y

CONFIG_SPI=y
CONFIG_LED_STRIP=y
CONFIG_LED_STRIP_LOG_LEVEL_DBG=y
CONFIG_WS2812_STRIP_SPI=y
CONFIG_WS2812_STRIP=y
```

https://pastebin.com/raw/4jgEm0RL

This looks to be the offending line...
```
CMake Warning at /home/machinehum/projects/id01-firmware/zephyr/CMakeLists.txt:809 (message):
  No SOURCES given to Zephyr library: drivers__ethernet

  Excluding target from build.
```

Any help would be appreciated! :)

fabiobaltieri commented:
@Machine-Hum nah that's just a warning, errors are the various:

```
/home/machinehum/.local/zephyr-sdk-0.15.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.1.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: cannot find -lnet80211: No such file or directory
/home/machinehum/.local/zephyr-sdk-0.15.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.1.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: cannot find -lcore: No such file or directory
...
```

I think you are missing the support libraries, have you ran `west blob fetch`?

Machine-Hum commented:
```
[machinehum@whitebox id01_fw]$ west blob fetch
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
west: error: argument <command>: invalid choice: 'blob' (choose from 'init', 'update', 'list', 'manifest', 'diff', 'status', 'forall', 'help', 'config', 'topdir', 'selfupdate', 'completion', 'boards', 'build', 'sign', 'flash', 'debug', 'debugserver', 'attach', 'zephyr-export', 'spdx', 'espressif')
```

My west is maybe old?
```
[machinehum@whitebox id01_fw]$ west --version
West version: v0.12.0
```

fabiobaltieri commented:
> [machinehum@whitebox id01_fw]$ west --version
> West version: v0.12.0
> ```

Sorry it's actually "blobs" (my bad), but yes I think you need to upgrade west (`pip3 install -U west`), current should be 0.14

Machine-Hum commented:
```
[machinehum@whitebox id01_fw]$ west blobs fetch
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
west: error: argument <command>: invalid choice: 'blobs' (choose from 'init', 'update', 'list', 'manifest', 'diff', 'status', 'forall', 'help', 'config', 'topdir', 'selfupdate', 'completion', 'boards', 'build', 'sign', 'flash', 'debug', 'debugserver', 'attach', 'zephyr-export', 'spdx', 'espressif')
[machinehum@whitebox id01_fw]$ west blob fetch
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
west: error: argument <command>: invalid choice: 'blob' (choose from 'init', 'update', 'list', 'manifest', 'diff', 'status', 'forall', 'help', 'config', 'topdir', 'selfupdate', 'completion', 'boards', 'build', 'sign', 'flash', 'debug', 'debugserver', 'attach', 'zephyr-export', 'spdx', 'espressif')
[machinehum@whitebox id01_fw]$ west --version
West version: v0.14.0
[machinehum@whitebox id01_fw]$ 
```
:(

fabiobaltieri commented:
> ```
> [machinehum@whitebox id01_fw]$ west blobs fetch
> usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
> west: error: argument <command>: invalid choice: 'blobs' (choose from 'init', 'update', 'list', 'manifest', 'diff', 'status', 'forall', 'help', 'config', 'topdir', 'selfupdate', 'completion', 'boards', 'build', 'sign', 'flash', 'debug', 'debugserver', 'attach', 'zephyr-export', 'spdx', 'espressif')
> [machinehum@whitebox id01_fw]$ west blob fetch
> usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
> west: error: argument <command>: invalid choice: 'blob' (choose from 'init', 'update', 'list', 'manifest', 'diff', 'status', 'forall', 'help', 'config', 'topdir', 'selfupdate', 'completion', 'boards', 'build', 'sign', 'flash', 'debug', 'debugserver', 'attach', 'zephyr-export', 'spdx', 'espressif')
> [machinehum@whitebox id01_fw]$ west --version
> West version: v0.14.0
> [machinehum@whitebox id01_fw]$ 
> ```
> 
> :(

Something's out of sync, can we follow up on a discussion https://github.com/zephyrproject-rtos/zephyr/discussions? Not worth keep adding on the PR page.

Machine-Hum commented:
https://github.com/zephyrproject-rtos/zephyr/discussions/50018
